powerauth-test-server/src/main/java/com/wultra/security/powerauth/app/testserver/service/ActivationService.java

```diff
@@ -0,0 +1,157 @@
+     * @throws GenericCryptographyException Thrown when cryptography computation fails.
+     */
+    @Transactional
+    @SuppressWarnings("unchecked")
+    public CreateActivationResponse createActivation(CreateActivationRequest request) throws AppConfigNotFoundException, GenericCryptographyException, RemoteExecutionException, ActivationFailedException {
+        // TODO - input validation
+        final Long applicationId = request.getApplicationId();
+        final Optional<TestConfigEntity> appConfigOptional = appConfigRepository.findById(applicationId);
+
+        if (appConfigOptional.isEmpty()) {
+            throw new AppConfigNotFoundException("Application configuration was not found for application ID: " + applicationId);
+        }
+
+        final TestConfigEntity appConfig = appConfigOptional.get();
+        final byte[] masterKeyBytes = BaseEncoding.base64().decode(appConfig.getMasterPublicKey());
+
+        final PublicKey publicKey;
+        try {
+            publicKey = keyConvertor.convertBytesToPublicKey(masterKeyBytes);
+        } catch (Exception ex) {
+            logger.warn("Key conversion failed, reason: {}", ex.getMessage());
+            logger.debug(ex.getMessage(), ex);
+            throw new GenericCryptographyException("Key conversion failed");
+        }
+
+        final JSONObject resultStatusObject = new JSONObject();
+
+        // Prepare activation
+        final PrepareActivationStepModel model = new PrepareActivationStepModel();
+        model.setActivationCode(request.getActivationCode());
+        model.setActivationName(request.getActivationName());
+        model.setApplicationKey(appConfig.getApplicationKey());
+        model.setApplicationSecret(appConfig.getApplicationSecret());
+        model.setMasterPublicKey(publicKey);
+        model.setHeaders(new HashMap<>());
+        model.setPassword(request.getPassword());
+        model.setResultStatusObject(resultStatusObject);
+        model.setUriString(config.getEnrollmentServiceUrl());
+        model.setVersion(config.getVersion());
+        model.setDeviceInfo("backend-tests");
+
+        String activationId = null;
+        try {
+            final ObjectStepLogger stepLogger = new ObjectStepLogger();
+            prepareActivationStep.execute(stepLogger, model.toMap());
+            for (StepItem item: stepLogger.getItems()) {
+                if ("Activation Done".equals(item.getName())) {
+                    final Map<String, Object> responseMap = (Map<String, Object>) item.getObject();
+                    activationId = (String) responseMap.get("activationId");
+                    break;
+                }
+            }
+        } catch (Exception ex) {
+            logger.warn("Remote execution failed, reason: {}", ex.getMessage());
+            logger.debug(ex.getMessage(), ex);
+            throw new RemoteExecutionException("Remote execution failed");
```

Comment:
This is a test tool, right? So there is no reason to hide the cause. I suggest to pass it to the exception constructor.

Reply:
Yes, we can do that.

Reply:
The component will be deployed in customer infrastructure for performance testing. On second thought, I suggest we keep the code as is to avoid leaking any sensitive information to potential attackers.